[warmup] Initial contribution

jamesmelville opened the pull request:
This binding integrates the Warmup 4iE Thermostat https://www.warmup.co.uk/thermostats/smart/4ie-underfloor-heating, via the API at https://my.warmup.com/.

I logged an issue for this at #7996. This PR was initially logged against 2.5.x in this PR #7998. All the comments in the PR have been addressed in this new PR for version 3.0.

Per the readme, it currently supports a Bridge to the API, holding the API credentials, and Rooms, which report current temperature and setpoint, and allow setting of an override of the setpoint for a specified duration.

I have tested this successfully on my own installation.

fwolter commented on bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/MyWarmupConfigurationDTO.java:
```diff
@@ -0,0 +1,25 @@
+ *
+ * @author James Melville - Initial contribution
+ */
+public class MyWarmupConfigurationDTO {
```
The configuration shouldn't be a DTO without NonNullByDefault annotation, as the fields can be null if the user uses textual configuration and forgets to specify a parameter.

fwolter commented on bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/RoomConfigurationDTO.java:
```diff
@@ -0,0 +1,32 @@
+ *
+ * @author James Melville - Initial contribution
+ */
+public class RoomConfigurationDTO {
```
See above.

fwolter commented on bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/RoomHandler.java:
```diff
@@ -0,0 +1,147 @@
+                        } else {
+                            updateStatus(ThingStatus.ONLINE);
+
+                            updateProperty("Id", room.getId());
+                            updateProperty("Serial Number", serialNumber);
+                            updateProperty("Name", room.getName());
+                            updateProperty("Location Id", location.getId());
+                            updateProperty("Location", location.getName());
```
As you use these properties in the discovery and below too, you could make the names constants.

fwolter commented on bundles/org.openhab.binding.warmup/src/main/resources/OH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,10 @@
+	xsi:schemaLocation="https://openhab.org/schemas/binding/v1.0.0 https://openhab.org/schemas/binding-1.0.0.xsd">
+
+	<name>Warmup Binding</name>
+	<description>This is the binding for a Warmup 4iE Thermostat normally used for controlling underfloor heating.</description>
```
```suggestion
	<description>This is the binding for a Warmup 4iE Thermostat primarily used for controlling underfloor heating.</description>
```

fwolter commented on bundles/org.openhab.binding.warmup/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,110 @@
+				<label>Serial Number</label>
+			</parameter>
+			<parameter name="overrideDurationMin" type="integer" unit="m" required="true">
+				<label>Override Duration (Minutes)</label>
```
The label should be concise. You could mention the unit in the description.

jamesmelville commented on bundles/org.openhab.binding.warmup/README.md:
```diff
@@ -0,0 +1,107 @@
+|--------------------|---------|-------------------------------------------------|----------|---------|
+| username           | String  | Username for my.warmup.com                      | true     |         |
+| password           | String  | Password for my.warmup.com                      | true     |         |
+| refreshIntervalSec | Integer | Interval in seconds between automatic refreshes | true     | 300     |
```
Hi @fwolter, thank you for the review and sorry for the slow response. In a previous review you suggested to add the unit to a property, but am I right in thinking that's only when the property is internal to the binding? https://github.com/openhab/openhab-addons/pull/7998#discussion_r460571680

fwolter replied:
That's right. Configuration parameter names shouldn't have the unit appended and class attributes should. I must admit that's a bit inconsistent. But it is what it is.

fwolter commented on bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/discovery/WarmupDiscoveryService.java:
```diff
@@ -0,0 +1,126 @@
+            if (localBridgeUID != null && deviceSN != null) {
+                final Map<String, Object> roomProperties = new HashMap<>();
+                roomProperties.put("serialNumber", deviceSN);
+                roomProperties.put("Serial Number", deviceSN);
```
I don't really understand the necessity for adding the serial number twice with different notations. Is it supposed to be a representation property and a human readable version?

jamesmelville replied:
exactly that. representation properties aren't displayed in paper UI after addition of a thing (although typically will be included in the thing id), so I thought it'd still be useful to display as a property.

fwolter replied:
OH3 shows also the property used as representation property in the UI.

jamesmelville replied:
Agreed - the idea was for convenience to have it on this page in paper UI for convenience
![image](https://user-images.githubusercontent.com/1859629/100921223-2695e180-34d4-11eb-8a95-739e4b05e326.png)
but I can see it's shown in this page twice so I'll remove it as I'm now only targeting v3
![image](https://user-images.githubusercontent.com/1859629/100921451-8096a700-34d4-11eb-86c5-0f916a9b03de.png)

jamesmelville commented:
Hi @fwolter thank you for your review - I've updated everything as requested except for the one conversation which remains open above. Alongside that I think I've simplified the error handling logic and removed some code which was stopping polling if an internet connection became unavailable, sorry if this adds to your work reviewing but I thought it was preferable. CI build hasn't completed but I'll keep an eye out for any issues raised. Also not sure how I added wborn as a code owner 🤷.

fwolter commented on bundles/org.openhab.binding.warmup/README.md:
```diff
@@ -0,0 +1,107 @@
+
+Once credentials are successfully added to the bridge, any rooms (devices) detected will be added as things to the inbox.
+
+## Thing Configuration
```
Can you mention the Thing Type IDs here?

fwolter commented on bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/WarmupBindingConstants.java:
```diff
@@ -0,0 +1,71 @@
+    public static final ThingTypeUID THING_TYPE_BRIDGE = new ThingTypeUID(BINDING_ID, "my-warmup");
+    public static final ThingTypeUID THING_TYPE_ROOM = new ThingTypeUID(BINDING_ID, "room");
+
+    public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Collections
+            .unmodifiableSet(new HashSet<>(Arrays.asList(THING_TYPE_BRIDGE, THING_TYPE_ROOM)));
+    public static final Set<ThingTypeUID> DISCOVERABLE_THING_TYPES_UIDS = Collections
+            .unmodifiableSet(new HashSet<>(Arrays.asList(THING_TYPE_ROOM)));
```
```suggestion
    public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Set.of(THING_TYPE_BRIDGE, THING_TYPE_ROOM);
    public static final Set<ThingTypeUID> DISCOVERABLE_THING_TYPES_UIDS = Set.of(THING_TYPE_ROOM);
```

fwolter commented on bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/discovery/WarmupDiscoveryService.java:
```diff
@@ -0,0 +1,125 @@
+        super(DISCOVERABLE_THING_TYPES_UIDS, 5, false);
+    }
+
+    @Override
+    public void activate() {
+        super.activate(null);
+    }
+
+    @Override
+    public void deactivate() {
+        super.deactivate();
+    }
```
Can these be removed?

jamesmelville replied:
I copied this pattern from this binding https://github.com/openhab/openhab-addons/blob/main/bundles/org.openhab.binding.draytonwiser/src/main/java/org/openhab/binding/draytonwiser/internal/discovery/DraytonWiserDiscoveryService.java, and I think this part of the code was rewritten by @Hilbrand (who also suggested previous changes to how I'd written the discovery service in #7998). I have to admit I'm not 100% up to speed with these services, so a bit of this was trial and error on my part.

It looks like I can remove the `activate()` method and the discovery still works but not the `deactivate()` method, which gives the compilation error `The inherited method AbstractDiscoveryService.deactivate() cannot hide the public abstract method in ThingHandlerService`, so I think I'm making the code compile (`deactivate()` in `AbstractDiscoveryService` is protected, but public in the `ThingHandlerService`), but because background discovery isn't enabled, the method might as well just be empty (and `super.activate(null)` wouldn't be doing anything either). Does this sound right?

fwolter replied:
That's fine.

fwolter commented on bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/RoomHandler.java:
```diff
@@ -0,0 +1,150 @@
+                    }
+                }
+            } catch (MyWarmupApiException e) {
+                logger.debug("Set Override failed: {}", e.getMessage(), e);
```
It seems a bit unreasonable to log the stack trace when the network fails. Same for below.

fwolter reviewed (changes requested):
The file `bundles/org.openhab.automation.groovyscripting/userdata/persist.mqtt` shouldn't be part of your PR.

What open point do you refer to?

jamesmelville commented:
> The file `bundles/org.openhab.automation.groovyscripting/userdata/persist.mqtt` shouldn't be part of your PR.
Sorry - I thought I'd already removed that but it's in my next push

> What open point do you refer to?
The duplicated serialNumber, which is now removed, so the point is no longer open.

fwolter commented on bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/MyWarmupAccountHandler.java:
```diff
@@ -0,0 +1,142 @@
+    }
+
+    public synchronized void refreshFromServer() {
+        try {
+            queryResponse = api.getStatus();
+        } catch (MyWarmupApiException e) {
+            queryResponse = null;
+            logger.debug("{}", e.getMessage(), e);
+        } finally {
+            if (queryResponse != null) {
+                updateStatus(ThingStatus.ONLINE);
+            } else {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "Unable to contact MyWarmup");
+            }
```
It seems a bit unreasonable to log the stack trace when the network fails. You could also include the exception's message in the ThingStatusDetailMessage.

jamesmelville replied:
Hi @fwolter , I've tweaked the error handling again, and made the changes to sync up with v3 changes - hope this is better!

fwolter reviewed (changes requested):
You need to replace `SmartHomeUnits` by `Units` to make the code compile.

fwolter commented on bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/MyWarmupAccountHandler.java:
```diff
@@ -95,15 +95,12 @@ public void cancelRefresh() {
         } catch (MyWarmupApiException e) {
             queryResponse = null;
-            logger.debug("{}", e.getMessage(), e);
+            logger.debug("{}", e.getMessage());
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
```
The log message could be removed, as the status update is already logged by the framework.

The state change originated by updateStatus() is logged to events.log. Including the status detail message.

fwolter reviewed (approved):
LGTM

cpmeister commented on bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/MyWarmupAccountHandler.java:
```diff
@@ -0,0 +1,135 @@
+        } catch (MyWarmupApiException e) {
+            queryResponse = null;
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+        } finally {
+            refreshFromCache();
+        }
```
Why is this in a finally clause? Please remove the finally clause if possible.

cpmeister commented on bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/RoomHandler.java:
```diff
@@ -0,0 +1,150 @@
+        QuantityType<?> temp = command.toUnit(SIUnits.CELSIUS);
+
+        if (temp != null) {
+            final int value = temp.multiply(new BigDecimal(10)).intValue();
```
```suggestion
            final int value = temp.multiply(BigDecimal.TEN).intValue();
```

cpmeister commented on bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/RoomHandler.java:
```diff
@@ -0,0 +1,150 @@
+    }
+
+    private void toggleFrostProtectionMode(OnOffType command) {
+        String roomId = getThing().getProperties().get("Id");
+        String locationId = getThing().getProperties().get("Location Id");
```
Thing properties should not have spaces in them. Please make the properties name camelcase if possible.

cpmeister commented on bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/WarmupThingHandler.java:
```diff
@@ -0,0 +1,93 @@
+    public void initialize() {
+        Bridge bridge = getBridge();
+        if (bridge != null) {
+            bridgeHandler = (MyWarmupAccountHandler) bridge.getHandler();
```
The bridge handler instance may change over the lifecycle of the thing handler. I suggest retrieving the bridge handler instance on demand instead of trying to cache it.

cpmeister commented:
Other than those few changes it looks good to me.

Hilbrand commented on bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/WarmupBindingConstants.java:
```diff
@@ -0,0 +1,66 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
```
The copyright year needs to be updated. You can use `mvn license:format -pl :org.openhab.binding.warmup` in the root. You need to rebase you binding to the current main to update (`git pull --rebase`) the configuration containing the new year before running this and than `git push --force-with-lease` update your remote branch,

Hilbrand reviewed (commented):
There are some last open review comments. And since we're already into th enew year, new release some basic must be updated. Are you able to finish the binding so it can be merged?

fwolter commented:
@jamesmelville can you fix the minor code review findings? Then we could merge this before OH 3.1 is released.

jamesmelville commented:
Hi @cpmeister @fwolter @Hilbrand - sorry for all the delays, life's been happening I guess! Anyway I think I've resolved all the review comments and got everything updated, not sure if I've done it in time for 3.1 but let me know!

fwolter reviewed (approved):
Thanks for addressing the remaining review comments!

LGTM

Your binding will make it into OH 3.1!

fwolter commented:
Now, you could add your binding's logo to the openHAB website. See https://deploy-preview-1540--openhab-docs-preview.netlify.app/docs/developer/bindings/#add-your-binding-s-logo-to-the-openhab-website
